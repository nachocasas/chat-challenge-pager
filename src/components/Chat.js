import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setUsername,
          handleConnectionStatusEvents, 
          setIncommingHistoryEvent,
          setIncommingUserlistEvent,
          setIncommingMessageEvent,
          setIncommingIsTypingEvent } from '../actions'
import io from "socket.io-client"

import History from './History';
import UserList from './UserList';
import ChatInput from './ChatInput';
import UsernamePropmtModal from './UsernamePropmtModal';
import { WhoIsTyping } from './WhoIsTyping'

const SOCKET_URL = 'http://localhost:3000'

class Chat extends Component {
  
  constructor(props){
    super(props);
    this.socket = null;
    this.setUserAndConnect = this.setUserAndConnect.bind(this);
    this.sendMessage = this.sendMessage.bind(this);
    this.setTyping = this.setTyping.bind(this);
  }

  setUserAndConnect(username){
    this.socket = io.connect(SOCKET_URL, { query: 'username='+username});
    this.props.setUsername(username);
    this.fireEventListeners();
  }

  sendMessage(message){
    this.socket.emit('message', message );
  }

  setTyping(isTyping){
    this.socket.emit('isTyping', isTyping );
  }

  whoIsTyping(){
    return this.props.isTyping ? this.props.isTyping.join(', ') : null
  }

  fireEventListeners(){
    const { props, socket } = this;
    props.handleConnectionStatusEvents(socket);
    props.setIncommingHistoryEvent(socket);
    props.setIncommingUserlistEvent(socket);
    props.setIncommingMessageEvent(socket);
    props.setIncommingIsTypingEvent(socket);
  }

  render(){
    const whoIsTyping = this.whoIsTyping();
    return(
      <div className="chat-app">
        { !this.props.connected && <UsernamePropmtModal onUsernameEnter = { this.setUserAndConnect } /> }
        <div className="header">
          <h1>Pager challenge chat</h1>
        </div>
        <div className="main-content">
          <div className="left-container">
            <History />
            <WhoIsTyping names={ whoIsTyping } />
            <ChatInput isTyping={ this.setTyping } onEnter={ this.sendMessage } />
          </div>
          <div className="right-container">
            <UserList onUserClick={ this.handleUserSelect } />
          </div>
        </div>
      </div>
    )
  }
}

function mapStateToProps({ chat }) {
  const { username, connected, isTyping } = chat;
  return { username, connected, isTyping }
}


export default connect(mapStateToProps, {
  setUsername,
  handleConnectionStatusEvents, 
  setIncommingHistoryEvent,
  setIncommingUserlistEvent,
  setIncommingMessageEvent, 
  setIncommingIsTypingEvent })(Chat)