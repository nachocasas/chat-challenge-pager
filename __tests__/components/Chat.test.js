import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import Chat from '../../src/components/Chat';

const mockStore = configureStore();

const INITIAL_STATE = 
{
  chat: {
    connected: true,
    username: 'Mike',
    history: [],
    userlist: ['Mike'],
    isTyping: []
    }
}

const store = mockStore(INITIAL_STATE);

describe('<Chat />', () => {
  describe('render()', () => {

    test('render component', () => {
      const wrapper = shallow(<Chat store={store} />);

      expect(toJson(wrapper)).toMatchSnapshot();
    });

  });
});