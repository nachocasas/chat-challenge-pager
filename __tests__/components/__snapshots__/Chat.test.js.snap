// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> render() render component 1`] = `
<Chat
  connected={true}
  handleConnectionStatusEvents={[Function]}
  isTyping={Array []}
  setIncommingHistoryEvent={[Function]}
  setIncommingIsTypingEvent={[Function]}
  setIncommingMessageEvent={[Function]}
  setIncommingUserlistEvent={[Function]}
  setUsername={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  username="Mike"
/>
`;
